packages/mdc-top-app-bar/standard/foundation.js

```diff
@@ -0,0 +1,183 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import MDCTopAppBarAdapter from '../adapter';
+import MDCTopAppBarBaseFoundation from '../foundation';
+import {numbers} from '../constants';
+
+/**
+ * @extends {MDCTopAppBarBaseFoundation<!MDCTopAppBarFoundation>}
+ * @final
+ */
+class MDCTopAppBarFoundation extends MDCTopAppBarBaseFoundation {
+  /**
+   * @param {!MDCTopAppBarAdapter} adapter
+   */
+  constructor(adapter) {
+    super(adapter);
+    // Used for diffs of current scroll position vs previous scroll position
+    /** @private {number} */
+    this.lastScrollPosition_ = this.adapter_.getViewportScrollY();
+    // Used to verify when the top app bar is completely showing or completely hidden
+    /** @private {number} */
+    this.topAppBarHeight_ = this.adapter_.getTopAppBarHeight();
+    // isDocked_ is used to indicate if the top app bar is 100% showing or hidden
+    /** @private {boolean} */
+    this.isDocked_ = true;
+    // Variable for current scroll position of the top app bar
+    /** @private {number} */
+    this.currentAppBarScrollPosition_ = 0;
+    // Used to prevent the top app bar from being scrolled out of view during resize events
+    /** @private {boolean} */
+    this.isCurrentlyBeingResized_ = false;
+    // The timeout that's used to throttle the resize events
+    /** @private {number} */
+    this.resizeThrottleId_ = -1;
+    // The timeout that's used to debounce toggling the isCurrentlyBeingResized_ variable after a resize
+    /** @private {number} */
+    this.resizeDebounceId_ = -1;
+
+    this.scrollHandler_ = () => this.topAppBarScrollHandler_();
+    this.resizeHandler_ = () => this.topAppBarResizeHandler_();
+  }
+
+  init() {
+    super.init();
+    this.adapter_.registerScrollHandler(this.scrollHandler_);
+    this.adapter_.registerResizeHandler(this.resizeHandler_);
+  }
+
+  destroy() {
+    super.destroy();
+    this.adapter_.deregisterScrollHandler(this.scrollHandler_);
+    this.adapter_.deregisterResizeHandler(this.resizeHandler_);
+    this.adapter_.setStyle('top', 'auto');
+  }
+
+  /**
+   * Function to determine if the DOM needs to update.
+   * @private
+   * @returns {boolean}
+   */
+  isUpdateRequired_() {
+    const partiallyShowing = this.currentAppBarScrollPosition_ < 0
+      && this.currentAppBarScrollPosition_ > -this.topAppBarHeight_;
+    let updateRequired = false;
+
+    if (this.isDocked_) {
+      // If it was previously already docked but now is partially showing, it's no longer docked.
+      if (partiallyShowing) {
+        this.isDocked_ = false;
```

Comment:
It feels a little weird to _set_ the value of an instance variable inside a _getter_ method.
It's not obvious that calling `isUpdateRequired_()` will have side effects.

Could we structure this a little differently and have `isDocked_` be a `get` property or something? So that it always returns the most up-to-date value (because it recomputes the value every time it's called).

Or, could we rename this method to something like `checkForUpdate_()`, so that it doesn't sound like a pure getter?

WDYT?